### ISSUE TITLE
Unable to duplicate BBF test accuracy

### ISSUE THREAD

**Issue by fsomers** [liuliu/ccv#61]
I'm new to _ccv_, so I thought I'd start by building latest stable (and unstable), and at least see if I could duplicate the **bbfdetect** face detection accuracy score from https://github.com/liuliu/ccv/blob/stable/doc/bbf.md

Unfortunately, I don't seem to be able to do so; I only get a score of **"53.82% (30)"** when I follow the steps; far worse than the quoted **"82.97% (12)"**.  I get the same result for stable and unstable branches.

I'm running on OSX 10.8.3 with the following _lib/config.mk_:

```
XCODE_SDK := macosx10.8
ARCH := x86_64
CC := clang
AR := ar
CFLAGS := -I/usr/local/include -arch x86_64 -isysroot /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.8.sdk -msse2 -D HAVE_SSE2 -D HAVE_LIBJPEG -D HAVE_LIBPNG -I/usr/X11/include -D HAVE_FFTW3 -D HAVE_LIBLINEAR -D HAVE_ACCELERATE_FRAMEWORK
LDFLAGS := -L/usr/local/lib -lm -arch x86_64 -isysroot /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.8.sdk -ljpeg -lpng -lz -L/usr/X11/lib -lfftw3f -lfftw3 -llinear -framework Accelerate
```

I have local installs in _/usr/local_ of _libpng_(1.5.14), _libjpeg_(v9), _fftw_(3.3.3), and _linear_(1.93).  I also tried _gsl_(1.15), with same results.  I have _XCode_(4.6.1) with up-to-date command line tools etc.

I converted the CMU _gif_ files to _png_ as per your example, using _ImageMagick_ (6.8.3-10, darwin 12.3.0). 

My generated CMU _filelist.txt_ has 130 entries, and _truth.txt_ has 511 entries.
After running **bbfdetect** on _filelist.txt_, the generated _result.txt_ has 305 entries.

Do you have any thoughts or suggestions as to how I might be able debug this further, so that I can match your results from the _bbf.md_ doc?

Thanks.

**Comment by liuliu** [comment]
my result doesn't includes the rotate face tests, make sure that is the case. I will re-check the result tonight.

**Comment by fsomers** [comment]
Thanks; I saw that the rotated cases were excluded in your bbf.md instructions. I prepared the data again to re-check, and I got the same results.

I have the newtest(65) + test(42) + test-low(23) = 130 input images in a single directory (and .png names in filelist.txt).  For truth test, I have Test Set A(169) + Test Set B(157) + Test Set C(185) = 511 entries in truth.txt.  All images freshly converted from gif to png using ImageMagick's 'convert' utility.

The CMU images.tar I downloaded has md5sum 9810f2660d4820cfe811c1600d95b898

**Comment by liuliu** [comment]
strange. I got 436 entries, and still can reprod the result presented in bbf.md doc with default bbfdetect.c and bbfvldr.rb script. But I am using Ubuntu with trunk version Clang. And yes, there are 511 entries in truth.txt and 130 images. Which validation script you use?

**Comment by fsomers** [comment]
I'm using un-modified bin/bbfdetect and bin/bbfvldtr.rb as per the doc - I can't find a bbfvldr.rb so I assumed it was a typo for the one I used (the same typo is in the bbf.md doc).

I'm running:

```
$ clang --version
Apple LLVM version 4.2 (clang-425.0.27) (based on LLVM 3.2svn)
Target: x86_64-apple-darwin12.3.0
Thread model: posix
```

Can you share the config.mk you are using, or perhaps cross-check the one I provided above?  I wasn't sure what the "minimum" set of options / libraries that I could get away with was, so I supplied as many of the options/ libs as I was readily able to.  Could you give me a "minimal" set of config.mk settings that should work?

Thanks.

**Comment by liuliu** [comment]
I think that you have the right configuration. Will try it on Mac OSX today.

**Comment by fsomers** [comment]
I wonder if you got a chance to try to reproduce this on OSX, or if you might have any suggestions for tests I might try?

**UPDATE**:  I just tried setting up a ubuntu VM (12.04-64bit with all updates applied), and reproduced the same "53.82% (30)" result as I had on OSX.  My ubuntu config is:

```
CC := clang
AR := ar
CFLAGS := -msse2 -D HAVE_SSE2 -D HAVE_LIBJPEG -D HAVE_LIBPNG
LDFLAGS := -lm -ljpeg -lpng -lz
```

I also tried dropping SSE2 support; same results.  I'm out of options, but I want to reproduce your results if I am to continue looking into ccv.

**Comment by liuliu** [comment]
@fsomers , my ground truth data is years old, will try to reprod again see if anything on that aspect changed.

**Comment by liuliu** [comment]
Hi, I just realized that we can cross check out ground truth data and ccv output data, here is mine:

Ground truth: https://gist.github.com/liuliu/5369039

ccv output: https://gist.github.com/liuliu/5369041

**Comment by fsomers** [comment]
Thanks for that.   My ground truth is the same, but my results/ output is different.  As a sanity check, I ran the bbfvldtr.rb script on your truth and output files, which gave me the "82.97% (12)" result, so the checker script is OK :)

Some possible causes come to mind, including:
- some silly mistake I've made that isn't yet obvious to either of us?
- some difference in conditional compilation?
- some difference in code generation?
- a fault with converting the images (ImageMagick related, or its usage by me?)
- a fault with loading the png images (some change in libpng between your ver, my ver?)

My copy of the images might help you check if there might be a fault/ difference with the image conversion / png library I'm using.  Might be worth checking to see if you can reproduce your results with my PNGs.

I've uploaded a tarball containing my PNG conversions of the CMU images, the truth file I made from the CMU data, my filelist and result output as produced by the commands below.  You can pull the archive (19MB) from dropbox at:
https://dl.dropboxusercontent.com/u/32394214/195ccv/fs-libccv-test-data.tar.gz
I'll delete the file after your next comment on this thread.

**UPDATED**: Upgraded to libpng 1.6.1 (no change in result)

In case of possible code differences, I started over with a fresh build of ccv-unstable (commit e7361af53c, grabbed 2013-04-11) on OSX.
Its a minimalist _config.mk_ - basically just _libpng_.  Note, I patched in _/usr/local/{include,lib}_ for my local _libpng_ (1.6.1 - March 28, 2013):

```
XCODE_SDK := macosx10.8
ARCH := x86_64
CC := clang
AR := ar
CFLAGS := -arch x86_64 -isysroot /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.8.sdk -D HAVE_LIBPNG -I/usr/local/include
LDFLAGS := -lm -arch x86_64 -isysroot /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.8.sdk -L/usr/local/lib -lpng -lz
```

In _bin/makefile_, I disabled building _icfcreate_ and _icfdetect_ temporarily so that I didn't require a dependency on _gsl_.

From _./bin_, I ran:

```
find ../images -name \*.png > ../fs-filelist.txt
./bbfdetect ../fs-filelist.txt ../samples/face > ../fs-results.txt
cat ../fs-results.txt | sed "s/\.\.\/images\///g" > ../fs-results-edited.txt
./bbfvldtr.rb ../fs-truth.txt ../fs-results-edited.txt
```

Result: "53.82% (30)"

You'll also find all of the .txt files mentioned in the commands above, in the archive on dropbox.  The commands assume you have expanded the archive, and placed the _images_ folder from it, one directory above _libccv/bin_.

Finally, just to confirm, I'm running:

```
uname -a
Darwin mc.local 12.3.0 Darwin Kernel Version 12.3.0: Sun Jan  6 22:37:10 PST 2013; root:xnu-2050.22.13~1/RELEASE_X86_64 x86_64
clang --version
Apple LLVM version 4.2 (clang-425.0.27) (based on LLVM 3.2svn)
Target: x86_64-apple-darwin12.3.0
Thread model: posix
```

Hope this helps.

**Comment by liuliu** [comment]
Hi, have you looked at the images you send to me? They are apparently incorrect: http://i.imgur.com/CxAFDe9.png notice that these thumbnails (and opened file) shouldn't be layered like this. Try another tool to do the image format conversion.

**Comment by fsomers** [comment]
I converted the images with ImageMagick (6.8.3-10) which was the latest version as of 2013-03-21, and I looked at a selection of them, but not all of them - I somehow managed to avoid the subset of images that have the artefact that you pointed out.

While investigating, I noticed that ImageMagick (6.8.3-10)'s _display_ program displays several of the un-converted _.gif_ files incorrectly, so that helped confirm the issue was likely with the GIF reader support in ImageMagick.  The artefact looks to me, like what I would expect to see if ImageMagick wasn't taking account of the interlace/ progressive image loading scheme in GIF - it must have been a recently introduced bug/ regression.

After some searching, I found an ImageMagick bug report relating to this issue here: http://www.imagemagick.org/discourse-server/viewtopic.php?f=3&t=23074&p=96642  which shows that the bug was only fixed on 2013-03-31, in ImageMagick (6.8.4-5).  So, I **updated to ImageMagick (6.8.4-8 2013-04-07) and the problem was resolved**.

I can now happily report that the images convert correctly with ImageMagick (6.8.4-8), and more importantly, my comparison run yielded the expected **"82.97% (12)"** to match your results.

Thank you for your suggestions, even though this eventually ended up being a (temporary) regression in ImageMagick.  A few weeks earlier, or later, and I wouldn't have encountered the issue.  Perhaps you could consider incorporating some self-tests in your project that might help others to validate their build of your code, and perhaps look to other components that might cause problems, sooner.

**Comment by liuliu** [comment]
Thanks for being patient. Yes, there are a lot of work ahead for ccv, but the point of my focus right now is still to have more the state of the art algorithms in the library (I am actually going to deprecate BBF as soon as the ICF one is ready, which expected to reach commercial grade face detector accuracy).
